sqlalchemy-stubs/orm/attributes.pyi

```diff
@@ -91,7 +92,10 @@ class Mapped(QueryableAttribute, Generic[_T]):
     def __get__(self, instance: None, owner: Any) -> "Mapped"[_T]: ...
     @overload
     def __get__(self, instance: object, owner: Any) -> _T: ...
+    @overload
     def __set__(self, instance: Any, value: _T) -> None: ...
+    @overload
+    def __set__(self, instance: Any, value: ClauseElement) -> None: ...
```

Comment:
This should be `ColumnElement[TypeEngine[_T]]`

Reply:
I tried this with #71, but this runs into errors for assigning `now()` as :
```
Incompatible types in assignment (expression has type "now[DateTime]", variable has type "ColumnElement[TypeEngine[Optional[datetime]]]")
```

The `a.integer = A.integer + 1` seems to crash mypy somewhere when inferring overloads, but I'm not sure if that is caused by #71 or this change.

Reply:
I think we are running into a covariance issue here, but I'm unsure how to solve it.

Given this test-case:

```py
from datetime import datetime

from sqlalchemy import Column, create_engine, DateTime, Integer
from sqlalchemy.orm import registry, Mapped, Session
from sqlalchemy.sql.functions import now

mapper_registry: registry = registry()
e = create_engine("sqlite:///database.db", echo=True)

@mapper_registry.mapped
class A:
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)
    date_time: Mapped[datetime] = Column(DateTime())
    date_time2 = Column(DateTime())

mapper_registry.metadata.create_all(e)

with Session(e) as s:
    a = A()
    a.date_time = now()
    a.date_time2 = now()
    s.add(a)
    s.commit()
```

I tried to adapt the `ColumnElement` type parameter to be covariant. It then correctly accepts `ColumnElement[TypeEngine[datetime]]` for `row[DateTime]`, however that does not solve the case `Mapped[Optional[datetime]]` case.

Reply:
I just tried your changes listed here rebased on top of the latest from `master` and the test case you have above passes. Do we still need the covariant solution?

Reply:
That is because the PR still contains the broad `ClauseElement`, instead of the `ColumnElement[TypeEngine[_T]]`. I didn't update the PR to be broken. Using `ColumnElement[TypeEngine[_T]]` as `value` results in:

```
mapped_assign_expression.py:21: error: Incompatible types in assignment (expression has type "now[DateTime]", variable has type "ColumnElement[TypeEngine[datetime]]")
mapped_assign_expression.py:22: error: Incompatible types in assignment (expression has type "now[DateTime]", variable has type "ColumnElement[TypeEngine[Optional[datetime]]]")
```

Reply:
Oh, I understand. I'll play with this today and see what I can come up with for a fix.

Reply:
As [discussed](https://gitter.im/sqlalchemy/devel/archives/2021/05/05) we will leave this `ClauseElement` for now and try to narrow it later.